iris-mpc-cpu/src/execution/hawk_main/matching.rs

```diff
@@ -179,8 +179,30 @@ impl BatchStep2 {
 pub struct BatchStep3(VecRequests<Step3>);
 
 impl BatchStep3 {
+    /// The final decision of whether the request is a match or unique.
+    ///
+    /// Emulate the behavior of inserting entries one by one. Intra-batch matches
+    /// only count if they are unique and being inserted themselves.
     pub fn is_matches(&self) -> VecRequests<bool> {
-        self.0.iter().map(Step3::is_match).collect_vec()
+        let mut is_matches = Vec::with_capacity(self.0.len());
+        for request in &self.0 {
+            let is_match = request
+                .select(Filter {
+                    eyes: Both,
+                    orient: Both,
+                    intra_batch: true,
+                })
+                .any(|id| match id {
+                    MatchId::Search(_) => true,
+                    MatchId::Luc(_) => true,
+                    MatchId::IntraBatch(request_i) => {
+                        let other_is_unique = !is_matches.get(request_i).unwrap_or(&true);
```

Comment:
I find this a bit confusing because you need to think about if the ! or the unwrap or is more strongly binding.
Maybe spelling out all 3 possibilities is easier to follow?:

Reply:
```suggestion
                        match is_matches.get(request_i) {
                            Some(true) => false, // the request we matched with was before us in the batch, and was a match, so we are not blocked by this intra-batch match
                            Some(false) => true, // the request we matched with was before us in the batch, and was unique, so we are blocked by this intra-batch match
                            None => false, // The request we matched with is after us in the batch, so we are not blocked by it.
                        }
```

Reply:
Thanks!